In [1]:
import sys
import time
import traceback

import models.mnist_data as mnist
from models.cnn_model import CNN
from modules.hpvconf import HPVGenerator
from modules.hpmgr import HPVManager
from modules.trainmgr import TrainingManager

In [2]:
LOG_PATH = "../log/learning_rate_test.csv"
TEMPLATE_PATH = 'CNN_HPV.ini'
CONFIG_PATH = 'config/'
NUM_GPUS = 1 # only one gpu can be used for HPOlib

In [15]:
def run(params):
    try:        
        generator = HPVGenerator()
        generator.setTemplate(TEMPLATE_PATH)
        hpv_file = generator.generate(params, output_dir=CONFIG_PATH)
                
        train_manager = TrainingManager(CNN(mnist.import_dataset()), LOG_PATH)
        train_manager.setTrainingDevices('gpu', NUM_GPUS)
        train_manager.setLoggingParams([ hyperparam for hyperparam in params])
        hpv = HPVManager(hpv_file, ini_dir=CONFIG_PATH)
        return train_manager.run(hpv, 3)
           
    except:
        e = sys.exc_info()
        print("Error: " + str(e))
        traceback.print_exc()    

In [4]:
use_decays = [True, False]
learning_rates = [0.1, 0.01, 0.0005]

for use_decay in use_decays:
    for learning_rate in learning_rates:
        params = {'USE_LEARNING_RATE_DECAY': use_decay, 'BASE_LEARNING_RATE': learning_rate}
        result = run(params)
        print str(params) + " -> " + str(result)

Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786308549326.ini is terminated properly
{'BASE_LEARNING_RATE': 0.1, 'USE_LEARNING_RATE_DECAY': True} -> 1.02
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786309595186.ini is terminated properly
{'BASE_LEARNING_RATE': 0.01, 'USE_LEARNING_RATE_DECAY': True} -> 0.63
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786310667985.ini is terminated properly
{'BASE_LEARNING_RATE': 0.0005, 'USE_LEARNING_RATE_DECAY': True} -> 1.51
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786311756898.ini is terminated properly
{'BASE_LEARNING_RATE': 0.1, 'USE_LEARNING_RATE_DECAY': False} -> 1.4
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786312787549.ini is terminated properly
{'BASE_LEARNING_RATE': 0.01, 'USE_LEARNING_RATE_DECAY': False} -> 0.74
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786313831394.ini is terminated properly
{'BASE_LEARNING_RATE': 0.0005, 'USE_LEARNING_RATE_DECAY': False} -> 1.29


In [5]:
use_decays = [True, False]
learning_rates = [0.005, 0.02, 0.05, 0.001, 0.0001]

for use_decay in use_decays:
    for learning_rate in learning_rates:
        params = {'USE_LEARNING_RATE_DECAY': use_decay, 'BASE_LEARNING_RATE': learning_rate}
        result = run(params)
        print str(params) + " -> " + str(result)

Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786315748617.ini is terminated properly
{'BASE_LEARNING_RATE': 0.005, 'USE_LEARNING_RATE_DECAY': True} -> 0.82
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786316795831.ini is terminated properly
{'BASE_LEARNING_RATE': 0.02, 'USE_LEARNING_RATE_DECAY': True} -> 0.6
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786317871794.ini is terminated properly
{'BASE_LEARNING_RATE': 0.05, 'USE_LEARNING_RATE_DECAY': True} -> 0.67
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786318938611.ini is terminated properly
{'BASE_LEARNING_RATE': 0.001, 'USE_LEARNING_RATE_DECAY': True} -> 1.23
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786320018503.ini is terminated properly
{'BASE_LEARNING_RATE': 0.0001, 'USE_LEARNING_RATE_DECAY': True} -> 3.56
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786321104892.ini is terminated properly
{'BASE_LEARNING_RATE': 0.005, 'USE_LEARNING_RATE_DECAY': False} -> 0.79
Training wi

* With learning rate decay
 * **{'BASE_LEARNING_RATE': 0.1}** -> 1.02
 * {'BASE_LEARNING_RATE': 0.05} -> 0.67
 * {'BASE_LEARNING_RATE': 0.02} -> 0.6
 * **{'BASE_LEARNING_RATE': 0.01}** -> 0.63
 * {'BASE_LEARNING_RATE': 0.005} -> 0.82
 * {'BASE_LEARNING_RATE': 0.001} -> 1.23
 * {'BASE_LEARNING_RATE': 0.0005} -> 1.51
 * **{'BASE_LEARNING_RATE': 0.0001}** -> 3.56
 

* Without learning rate decay
 * **{'BASE_LEARNING_RATE': 0.1}** -> 1.4
 * {'BASE_LEARNING_RATE': 0.05} -> 0.85
 * {'BASE_LEARNING_RATE': 0.02} -> 0.77
 * **{'BASE_LEARNING_RATE': 0.01}** -> 0.74
 * {'BASE_LEARNING_RATE': 0.005} -> 0.79
 * {'BASE_LEARNING_RATE': 0.001} -> 1.14
 * {'BASE_LEARNING_RATE': 0.0005} -> 1.29
 * **{'BASE_LEARNING_RATE': 0.0001}** -> 2.91
 

- tested learning rate: 0.1, 0.05, 0.02, 0.01, 0.05, 0.001, 0.0005, 0.0001 (w/ and w/o learning rate decay (0.95))
  * test error decreased near 0.02~0.01 and increased again 
    * test errors w/ learning late decay : 1.02, 0.67, 0.6, 0.63, 0.82,  1.23, 1.51, 3.56
    * test errors w/o  learning late decay : 1.4, 0.85, 0.77, 0.74, 0.79, 1.14, 1.29, 2.91

In [6]:
LOG_PATH = "../log/regularization_test.csv"

In [8]:
regularization_factors = [0, 5e-5, 1e-4, 2e-4, 3e-4, 4e-4, 6e-4, 7e-4, 8e-4, 9e-4, 5e-3, 5e-2]

for regularization_factor in regularization_factors:
    params = {'REGULARIZER_FACTOR': regularization_factor}
    result = run(params)
    print str(params) + " -> " + str(result) + "%"

Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786746824296.ini is terminated properly
{'REGULARIZER_FACTOR': 0} -> 0.6%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786747884366.ini is terminated properly
{'REGULARIZER_FACTOR': 5e-05} -> 0.59%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786748958402.ini is terminated properly
{'REGULARIZER_FACTOR': 0.0001} -> 0.64%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786750034210.ini is terminated properly
{'REGULARIZER_FACTOR': 0.0002} -> 0.63%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786751109682.ini is terminated properly
{'REGULARIZER_FACTOR': 0.0003} -> 0.69%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786752184817.ini is terminated properly
{'REGULARIZER_FACTOR': 0.0004} -> 0.67%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786753260488.ini is terminated properly
{'REGULARIZER_FACTOR': 0.0006} -> 0.67%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786754353644.in

In [9]:
regularization_factors = [5e-8, 5e-7, 5e-6, 5e-1, 4e-1, 3e-1, 2e-1, 1e-1]

for regularization_factor in regularization_factors:
    params = {'REGULARIZER_FACTOR': regularization_factor}
    result = run(params)
    print str(params) + " -> " + str(result) + "%"

Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786787480803.ini is terminated properly
{'REGULARIZER_FACTOR': 5e-08} -> 0.67%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786788520223.ini is terminated properly
{'REGULARIZER_FACTOR': 5e-07} -> 0.67%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786789605891.ini is terminated properly
{'REGULARIZER_FACTOR': 5e-06} -> 0.7%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786790696790.ini is terminated properly
{'REGULARIZER_FACTOR': 0.5} -> 4.32%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786791791477.ini is terminated properly
{'REGULARIZER_FACTOR': 0.4} -> 3.16%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786792882536.ini is terminated properly
{'REGULARIZER_FACTOR': 0.3} -> 2.97%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786793987900.ini is terminated properly
{'REGULARIZER_FACTOR': 0.2} -> 2.64%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786795079631.ini is term

* {'REGULARIZER_FACTOR': 0} -> 0.6%
* {'REGULARIZER_FACTOR': 5e-08} -> 0.67%
* {'REGULARIZER_FACTOR': 5e-07} -> 0.67%
* {'REGULARIZER_FACTOR': 5e-06} -> 0.7%
* {'REGULARIZER_FACTOR': 5e-05} -> 0.59%
* {'REGULARIZER_FACTOR': 0.0001} -> 0.64%
* {'REGULARIZER_FACTOR': 0.0002} -> 0.63%
* {'REGULARIZER_FACTOR': 0.0003} -> 0.69%
* {'REGULARIZER_FACTOR': 0.0004} -> 0.67%
* {'REGULARIZER_FACTOR': 0.0006} -> 0.67%
* {'REGULARIZER_FACTOR': 0.0007} -> 0.73%
* {'REGULARIZER_FACTOR': 0.0008} -> 0.68%
* {'REGULARIZER_FACTOR': 0.005} -> 0.99%
* {'REGULARIZER_FACTOR': 0.05} -> 2.37%
* {'REGULARIZER_FACTOR': 0.1} -> 2.12%
* {'REGULARIZER_FACTOR': 0.2} -> 2.64%
* {'REGULARIZER_FACTOR': 0.3} -> 2.97%
* {'REGULARIZER_FACTOR': 0.4} -> 3.16%
* {'REGULARIZER_FACTOR': 0.5} -> 4.32%

After regularizer factor testing, followings may be effective:
 * 0 ~ 0.001 : 0.6? % -> choose **5e-05**
 * 0.005 ~ 0.05 : 0.99% ~ 2.37% -> choose **0.25**
 * other : choose **0.5**  

In [10]:
LOG_PATH = "../log/dropout_test.csv"

In [13]:
dropouts = [0.9, 0.8, 0.75, 0.7, 0.6, 0.4, 0.3, 0.25, 0.2, 0.1]

for dropout in dropouts:
    params = {'DROPOUT_RATE': dropout}
    result = run(params)
    print str(params) + " -> " + str(result) + "%"

Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786818570633.ini is terminated properly
{'DROPOUT_RATE': 0.9} -> 0.77%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786819626294.ini is terminated properly
{'DROPOUT_RATE': 0.8} -> 0.77%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786820733148.ini is terminated properly
{'DROPOUT_RATE': 0.75} -> 0.73%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786821833540.ini is terminated properly
{'DROPOUT_RATE': 0.7} -> 0.69%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786822936462.ini is terminated properly
{'DROPOUT_RATE': 0.6} -> 0.75%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786824038998.ini is terminated properly
{'DROPOUT_RATE': 0.4} -> 0.68%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786825145771.ini is terminated properly
{'DROPOUT_RATE': 0.3} -> 0.75%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786826249595.ini is terminated properly
{'DROPOUT_RATE': 0.25} -> 0.58

Traceback (most recent call last):
  File "models/cnn_model.py", line 79, in learn
    y = self.train()
  File "models/cnn_model.py", line 107, in train
    logits = self.leNet5(self.train_data_node, True)
  File "models/cnn_model.py", line 392, in leNet5
    hidden = tf.nn.dropout(hidden, 1.0 - self.hp["DROPOUT_RATE"], seed=self.hp["SEED"])
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/nn_ops.py", line 1114, in dropout
    "range (0, 1], got %g" % keep_prob)
ValueError: keep_prob must be a scalar tensor or a float in the range (0, 1], got 0


following dropout rates are selected: **1.0, 0.75, 0.25**

In [18]:
LOG_PATH = "../log/reg_mix_test.csv"

In [16]:
use_decays = [True, False]
learning_rates = [0.1, 0.01, 0.0001]
regularization_factors = [5e-05, 0.25, 0.5]
dropouts = [1., 0.75, 0.25]
for use_decay in use_decays:
    for learning_rate in learning_rates:
        for regularization_factor in regularization_factors:
            for dropout in dropouts:    
                params = {'USE_LEARNING_RATE_DECAY': use_decay, 'BASE_LEARNING_RATE': learning_rate, \
                      'REGULARIZER_FACTOR': regularization_factor, 'DROPOUT_RATE': dropout}
                result = run(params)
                print str(params) + " -> " + str(result) + "%"

Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786915277472.ini is terminated properly
{'DROPOUT_RATE': 1.0, 'BASE_LEARNING_RATE': 0.1, 'REGULARIZER_FACTOR': 5e-05, 'USE_LEARNING_RATE_DECAY': True} -> 1.71%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786916317935.ini is terminated properly
{'DROPOUT_RATE': 0.75, 'BASE_LEARNING_RATE': 0.1, 'REGULARIZER_FACTOR': 5e-05, 'USE_LEARNING_RATE_DECAY': True} -> 89.72%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786917411138.ini is terminated properly
{'DROPOUT_RATE': 0.25, 'BASE_LEARNING_RATE': 0.1, 'REGULARIZER_FACTOR': 5e-05, 'USE_LEARNING_RATE_DECAY': True} -> 89.72%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786918492949.ini is terminated properly
{'DROPOUT_RATE': 1.0, 'BASE_LEARNING_RATE': 0.1, 'REGULARIZER_FACTOR': 0.25, 'USE_LEARNING_RATE_DECAY': True} -> 90.2%
Training with /root/tf-hpolib/tf-hpo/src/config/HPV_14786919573739.ini is terminated properly
{'DROPOUT_RATE': 0.75, 'BASE_LEARNING_RATE': 0.

In [ ]:
#TODO: consider other combination